fontbe/src/hvar.rs

```diff
@@ -0,0 +1,191 @@
+                // therefore the deltas returned from VariationModel will be an empty Vec.
+                // However, when this is the first .notdef glyph we would like to treat it
+                // specially in order to match the output of fontTools.varLib.
+                // In fonttools, all master TTFs have a .notdef glyph as their first glyph; in fontc,
+                // unless the input source defines a .notdef, only a default instance is generated.
+                // And that's ok for gvar, however for HVAR the order in which regions and associated
+                // deltas are added to VariationStoreBuilder, one glyph at a time, can produce
+                // different orderings of the ItemVariationStore.VariationRegionList (newly seen
+                // regions get appended, and existing regions reused).
+                // So, to match the VarRegionList produced by fontTools, we need to make the deltaset
+                // for the first .notdef glyph similarly "dense", by copying its default instance to
+                // all other glyph locations...
+                if i == 0 && gn.as_str() == ".notdef" {
+                    let notdef_width = advance_widths.values().next().unwrap()[0];
+                    for loc in global_locations.iter() {
+                        advance_widths
+                            .entry(loc.clone())
+                            .or_insert_with(|| vec![notdef_width]);
+                    }
+                } else {
+                    // spare the model the work of computing no-op deltas
+                    all_glyph_width_deltas.push(Vec::new());
+                    continue;
+                }
+            }
+            let locations = advance_widths.keys().cloned().collect::<BTreeSet<_>>();
+            let model = models.entry(locations).or_insert_with(|| {
+                // this glyph defines its own set of locations, a new sparse model is needed
+                single_model = false;
+                VariationModel::new(advance_widths.keys().cloned().collect(), axes.clone()).unwrap()
+            });
+            all_glyph_width_deltas.push(
+                model
+                    .deltas(&advance_widths)
+                    .map_err(|e| Error::GlyphDeltaError(gn.clone(), e))?
+                    .into_iter()
+                    .filter_map(|(region, values)| {
+                        if region.is_default() {
+                            return None;
+                        }
+                        // Only 1 value per region for our input
+                        assert!(values.len() == 1, "{} values?!", values.len());
+                        Some((
+                            region.to_write_fonts_variation_region(axes),
+                            values[0].ot_round(),
+                        ))
+                    })
+                    .collect(),
+            )
+        }
```

Comment:
I spent a reasonable amount of time staring at this loop trying to figure out what was happening and why. I don't see anything wrong with it but I think it might be possible to make easier on the reader.

My immediate thought is to make `all_glyph_width_deltas: Vec<Vec<(VariationRegion, i16)>>` a struct and then called a method on it to add glyphs, e.g.

```rust
        let mut deltas = Deltas::new();
        for (i, gn) in glyph_order.iter().enumerate() {
            let glyph = context.ir.glyphs.get(&FeWorkId::Glyph(gn.clone()));
            deltas.add(glyph.as_ref());
        }
```

Deltas would own the models map. The single_model bool is replaced by Deltas::is_single_model. `exec` gets a bit closer to being a short method you can read quickly.

Reply:
yeah, that for loop is pretty big/ugly, I will try to refactor like you suggest next week, thanks

Reply:
Done in https://github.com/googlefonts/fontc/pull/490/commits/3f5e6f8092918929382344a25fd8b72002e8c766